### PR TITLE
gdal: fix std::shared_mutex build failure; enable build with libxml2 2.12

### PR DESCRIPTION
**Pull request by nilason**
#### Description

- fix build failure on missing std::shared_mutex
   Closes https://trac.macports.org/ticket/68716
- enable build with libxml2 version 2.12
   (suggested in https://lists.osgeo.org/pipermail/gdal-dev/2023-November/057965.html)
-  set CXX standard to 2017 for macOS 10.12+
   enables the use of std::shared_mutex, which was introduced with version 3.8.0,
for systems that support it.

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 13.6.2 22G320 arm64
Xcode 15.0.1 15A507

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: gis/gdal/Portfile [hunk 1]
```diff
@@ -36,7 +36,13 @@ use_xz              yes
...
 compiler.blacklist-append {clang < 900}
 
-compiler.cxx_standard 2014
+# Enable use of std::shared_mutex on systems that support it (macOS 10.12+),
+# next major GDAL version (>3.8) will require C++17.
+if {${os.platform} eq "darwin" && ${os.major} < 16} {
+    compiler.cxx_standard 2014
+} else {
+    compiler.cxx_standard 2017
+}
```
Comment (reneeotten) [inline gis/gdal/Portfile:45]: why can you not set the CXX to 2017 on all systems? It should in all cases select a compiler that support the "shared mutex" thing; either by an Apple provided Clang or a MacPorts installed one.

Comment (kencu) [reply gis/gdal/Portfile:45]: the c++17 and newer features were gradually introduced to libc++.dylb.

so only the /usr/lib/libc++.dylib in 10.12 and newer have it.

```
define _LIBCPP_AVAILABILITY_SHARED_MUTEX                                    \
         __attribute__((availability(macosx,strict,introduced=10.12)))           \
```

Comment (nilason) [reply gis/gdal/Portfile:45]: Thanks @kencu, that confirms what I deducted from the build bots.

@reneeotten GDAL still only requires C++11 (don't know why the Portfile bumped to C++14 with the deletion of the C++11 PG, perhaps because of some dependency, never mind). Only the next major GDAL release (probably 4.0) will require C++17 (the cmake change is already in master). The developers decided to optionally insert C++17 feats, like std::shared_mutex, the included patch improves the detection of its availability.
To sum up: there is **now** no compelling reason to make this C++17 all over and in the process perhaps run into new unexpected issues...
